exonum/src/helpers/fabric/password.rs

```diff
@@ -0,0 +1,164 @@
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This module contains utilities for passphrase entry.
+
+use failure;
+use rpassword::read_password_from_tty;
+
+use std::{env, io, str::FromStr};
+
+use helpers::ZeroizeOnDrop;
+
+/// Passphrase input methods
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub enum PassInputMethod {
+    /// Prompt passphrase from terminal.
+    Terminal,
+    /// Get passphrase from environment variable (default if `None`).
+    EnvVariable(Option<String>),
+    /// Passphrase is passed as a command line parameter.
+    CmdLineParameter(ZeroizeOnDrop<String>),
+}
+
+/// Secret key types.
+#[derive(Copy, Clone)]
+pub enum SecretKeyType {
+    Consensus,
+    Service,
+}
+
+impl PassInputMethod {
+    /// Get passphrase using selected method.
+    /// Details of this process differs for different secret key types and whether we run node
+    /// or generate config files.
+    pub fn get_passphrase(self, key_type: SecretKeyType, node_run: bool) -> ZeroizeOnDrop<String> {
+        match self {
+            PassInputMethod::Terminal => {
+                let prompt = match key_type {
+                    SecretKeyType::Consensus => "Enter consensus key passphrase",
+                    SecretKeyType::Service => "Enter service key passphrase",
+                };
+                prompt_passphrase(prompt, node_run).expect("Failed to read password from stdin")
+            }
+            PassInputMethod::EnvVariable(name) => {
+                let var = if let Some(ref name) = name {
+                    name
+                } else {
+                    match key_type {
+                        SecretKeyType::Consensus => "EXONUM_CONSENSUS_PASS",
+                        SecretKeyType::Service => "EXONUM_SERVICE_PASS",
+                    }
+                };
+                ZeroizeOnDrop(env::var(var).unwrap_or_else(|e| {
+                    panic!("Failed to get password from env variable: {}, {}", var, e)
+                }))
+            }
+            PassInputMethod::CmdLineParameter(pass) => pass,
+        }
+    }
+}
+
+impl Default for PassInputMethod {
+    fn default() -> Self {
+        PassInputMethod::Terminal
+    }
+}
+
+impl FromStr for PassInputMethod {
+    type Err = failure::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == "" {
+            return Ok(Default::default());
+        }
+
+        if s == "stdin" {
+            return Ok(PassInputMethod::Terminal);
+        }
+
+        if s.starts_with("env") {
+            let env_var = s.split(':').nth(1).map(|s| s.to_owned());
```

Comment:
`map(String::to_owned)` should works too.

Reply:
I always get this kind of comment and I do not see any benefit.
If you all really prefer it you should add it to CONTRIBUTIND.md :)

Reply:
Also it does not work. It should be`str::to_owned` or something like that.

Reply:
FYI, https://rust-lang.github.io/rust-clippy/master/#redundant_closure

Reply:
However, I'm not insisting on this specific case. If clippy is satisfied - leave it as it is